### ISSUE TITLE
[bug]: TLS connection cuts with "sslv3 alert bad record mac" when sending a lot of outbound requests

### ISSUE THREAD

**Issue by escatg** [asterisk/asterisk#168]
### Severity

Minor

### Versions

18.16

### Components/Modules

pjproject, res_pjsip

### Operating Environment

Ubuntu  22.04 (libssl3)
Ubuntu 20.04 (libssl1.1)

### Frequency of Occurrence

Frequent

### Issue Description

Using outbound SIP TLS registrations with res_pjsip, we noticed that whenever the system was under load and a lot of requests transited using the same TLS connection, we would often have connection cuts. The remote end would close the connection sending a "bad record mac" alert back to asterisk.

Since this would only happen under load and after some time, I tried to make it easy to reproduce. I created a module that sends a lot of SIP OPTIONS in parallel to an endpoint and using it I could reproduce it pretty quickly. To use it, you need to have one endpoint that uses a TLS transport and run "pjsip floodping <endpoint>". I usually put sipp in front as a responder.

Creating an issue here since I'm not sure if it is a problem in pjproject itself or in the way asterisk uses it.

[res_pjsip_ping.c.txt](https://github.com/asterisk/asterisk/files/11799828/res_pjsip_ping.c.txt)

### Relevant log output

```shell
WARNING[3966790] pjproject: 	                   SSL SSL_ERROR_SSL (Read): Level: 0 err: <336151548> <SSL routines-ssl3_read_bytes-sslv3 alert bad record mac> len: 65535 peer: 172.16.33.79:5067

[tlstrans]
type=transport
protocol=tls
bind=172.16.33.79:5061
method=tlsv1_2

[sipp]
type=endpoint
context=FromEPA
disallow=all
allow=ulaw
aors=sipp
transport=tlstrans
 
[sipp]
type=aor
contact=sip:172.16.33.79:5067
qualify_frequency=60
```

### Asterisk Issue Guidelines

- [X] Yes, I have read the Asterisk Issue Guidelines

**Comment by jcolp** [comment]
We don't do anything special for transports really. We send requests to an IP address and port using a specific transport type, and PJSIP does the rest so this is most likely an issue in PJSIP itself - including doing connection reuse.

**Comment by jcolp** [comment]
@creslin287 With your recent foray into TLS, have you seen anything like this?

**Comment by creslin287** [comment]
I'm currently looking into some issues on our side that could possibly be symptoms of the same underlying bug, but not sure yet.

**Comment by escatg** [comment]
From my tests, it seems the trigger is having a lot of sip worker threads trying to send requests at the same time. Indeed like you said this is totally offloaded to PJSIP and from what I could see it uses locking to prevent simultaneous writes to the ssl socket (third-party/pjproject/source/pjlib/src/pj/ssl_sock_imp_common.c:1703), so I wonder what is happening.

**Comment by creslin287** [comment]
Can you test this patch out to see if it makes a difference (it's against pjproject, so you'll have to apply it there)?  I just fixed an issue around this and it was caused by connections leaving stale errors in the thread local error queue:

```
diff --git a/pjlib/src/pj/ssl_sock_ossl.c b/pjlib/src/pj/ssl_sock_ossl.c
index d31cd5c82..ec2dad5fe 100644
--- a/pjlib/src/pj/ssl_sock_ossl.c
+++ b/pjlib/src/pj/ssl_sock_ossl.c
@@ -2361,6 +2361,10 @@ static pj_status_t ssl_do_handshake(pj_ssl_sock_t *ssock)
 
     /* Perform SSL handshake */
     pj_lock_acquire(ssock->write_mutex);
+
+    /* Clear the error queue prior to any I/O functions, as per openssl docs */
+    ERR_clear_error();
+
     err = SSL_do_handshake(ossock->ossl_ssl);
     pj_lock_release(ssock->write_mutex);
 
@@ -2444,6 +2448,10 @@ static pj_status_t ssl_read(pj_ssl_sock_t *ssock, void *data, int *size)
      * is on progress, so let's protect it with write mutex.
      */
     pj_lock_acquire(ssock->write_mutex);
+
+    /* Clear the error queue prior to any I/O functions, as per openssl docs */
+    ERR_clear_error();
+
     *size = size_ = SSL_read(ossock->ossl_ssl, data, size_);
 
     if (size_ <= 0) {
@@ -2494,6 +2502,9 @@ static pj_status_t ssl_write(pj_ssl_sock_t *ssock, const void *data,
     ossl_sock_t *ossock = (ossl_sock_t *)ssock;
     pj_status_t status = PJ_SUCCESS;
 
+    /* Clear the error queue prior to any I/O functions, as per openssl docs */
+    ERR_clear_error();
+
     *nwritten = SSL_write(ossock->ossl_ssl, data, (int)size);
     if (*nwritten <= 0) {
         /* SSL failed to process the data, it may just that re-negotiation
@@ -2528,6 +2539,9 @@ static pj_status_t ssl_renegotiate(pj_ssl_sock_t *ssock)
     if (SSL_renegotiate_pending(ossock->ossl_ssl))
         return PJ_EPENDING;
 
+    /* Clear the error queue prior to any I/O functions, as per openssl docs */
+    ERR_clear_error();
+
     ret = SSL_renegotiate(ossock->ossl_ssl);
     if (ret <= 0) {
         status = GET_SSL_STATUS(ssock);
```

**Comment by jcolp** [comment]
@escatg Have you been able to try the patch provided by Matt?

**Comment by creslin287** [comment]
@escatg bump bump.  Any luck?

**Comment by escatg** [comment]
Sorry for the late reply, I somehow missed the notifications, fixed that now.

I applied the patch (on a newer asterisk 18.19 this time) and tested, but unfortunately this doesn't change anything.

Looking at the patch it is not surprising as it clears errors that might be left in the queue, sure, but the "sslv3 alert bad record mac" comes from the remote end on the next read.

**Comment by jcolp** [comment]
@creslin287 Now that we've received confirmation the patch didn't resolve it, have you seen any further issues under load in your own TLS based environment?

**Comment by creslin287** [comment]
I'd say that we still have some small percentage of ongoing SSL_ERROR_SSLs still, but nothing at this point that has been a strong enough signal or large enough problem to dig deeper into (assumption at this point is that they're one offs, misconfigured endpoints, attackers, etc).

I wouldn't be surprised that there might be a few more bugs in PJPROJECT's TLS stack that need to be found though (lots of funny code paths on pjproject that I still am not sure how some of them are safe).

**Comment by escatg** [comment]
Since pjproject was upgraded to 2.14.1 in latest asterisk, I tested again. I can still reproduce this TLS connection cut with the code I provided earlier.

**Comment by hayzamjs** [comment]
I am hitting the same issue with a lot of phones (remote (around 30) and otherwise (also around 30)) with presence, TLS enabled for all of them, I will try and get you guys a config that can make it easier to reproduce perhaps.

**Comment by p-barabas** [comment]
Hello, i have the same issue. I refresh registrations from kamailio to asterisk periodically (150-200 registrations) in a loop, and sometimes I got this SSL_ERROR_SSL error which removes all contacts, so every one should be registered again. Any idea or solution for it?

**Comment by escatg** [comment]
I just reproduced this using asterisk 22.1 just by loading it with the res_pjsip_ping module provided, running on a recent platform with up-to-date libssl. Updated the description accordingly.

I will retry with master to see if it is still the case with pjproject 2.15.1.

**Comment by escatg** [comment]
I can also reproduce it on 4ef68fbce8df8b63759995e2cddc2331ec344b79

**Comment by hjelmeland** [comment]
Chime in to tell we have also been bitten by the TLS issue. Asterisk 22.1 with bundled pjproject. Using TLS 1.2.

Scenario: We have stitched together a call queueing application with priority handling using dialplan and database programming. When a high priority call arrives on top of a number of low priory calls already queued to operator phones, dialplan run a loop to cancel the low priory calls before sending the high priority call to operator phones. With a looped QA soak test script this fails too often for comfort. 

When it fails, we see the outgoing SIP TLS connection to operator phone was disconnected for a moment.

Forcing cipher to 0x009D (TLS_RSA_WITH_AES_256_GCM_SHA384) so we could decrypt a wireshark trace, we see that Asterisk sends a TLS frame (containing a SIP CANCEL) with incorrect MAC. Phone respond with a TLS Alert frame with level _fatal_, making Asterisk hang up the TLS connection.

Suspecting lack of locking somewhere: As a work around we added a Wait(0.1) to the cancel loop. That decreased the failure rate by more than factor 10.

I know next to nothing about Asterisk, pjlib and openssl internals, so this is just speculation:
Looking in pjproject pjlib/src/pj/ssl_sock_ossl.c I see no locking except locking based on CRYPTO_set_locking_callback().
CRYPTO_set_locking_callback() is just a void macro now, the CRYPTO_set_locking_callback code was removed back in openssl 1.1.1 in 2016. Looks suspicious to me, but as said, I don't understand the design.

**Comment by kobaz** [comment]
I can confirm that the bad mac problem exists even without a high load.

I have a site with 33 TLS endpoints and they just randomly go unreachable/reachable.  Every single time this occurs the error is
` <SSL routines-ssl3_read_bytes-sslv3 alert bad record mac>`

**Comment by jacotec** [comment]
What phones are you guys using? I'm suffering the same issue, but only with my Yealink phones (and the Groundwire app, but that can be due to network changes).

Auerswald D-Series and some ATA behave well and did not cause this a single time.

**Comment by intellasoft** [comment]
We're seeing this on Yealinks for sure.  TLS Is almost unusable. We've been switching back to TCP and UDP because phones go offline regularly due to this issue.

**Comment by azertyfun** [comment]
Applying https://github.com/pjsip/pjproject/pull/4543 fixes the issue reproducible with the `floodping` module provided by @escatg. It was initially introduced by https://github.com/pjsip/pjproject/issues/1573.

It's such an obvious issue ([intentionally breaking OpenSSL's thread-safety paradigm](https://github.com/pjsip/pjproject/blob/2.15.1/pjlib/src/pj/ssl_sock_imp_common.c#L1829)) that I initially discarded it as "surely whoever wrote that knew what they were doing".

The fix was made available with pjproject 2.16: https://github.com/pjsip/pjproject/commit/c63b568be3880b19c20ed4b7a86dbf0406fe3cd1

**Comment by seanbright** [comment]
Closed by #1747

**Comment by hjelmeland** [comment]
https://github.com/asterisk/asterisk/pull/1747 has not been cherry-picked to releases/22. Will it? Just so we know whether to wait for 22.8.3, or patch locally.

**Comment by jcolp** [comment]
As you can see in the PR:

Successfully merged to branch master and cherry-picked to ["20","22","23"]

It was merged to all of the currently supported branches, it will therefore be in the next set of releases. It will not trigger a 22.8.3

**Comment by azertyfun** [comment]
Turns out I was a bit too excited to share my discovery. The bugfix referenced (https://github.com/pjsip/pjproject/pull/4543) makes things **worse**. When multiple threads are concurrently sending packets, PJSIP may instead deadlock, which unlike with the pre-fix behavior completely blocks the SIP stack instead of "merely" causing a transport re-negotiation.

I identified the deadlock in https://github.com/pjsip/pjproject/issues/4818. Unfortunately no obvious fix came to me; hopefully a PJSIP maintainer can help me out with this.

**Comment by azertyfun** [comment]
Update: Upstream considers that the only fix is to enable `PJ_IOQUEUE_CALLBACK_NO_LOCK` in pjproject. I tend to agree; the fundamental architecture of the TLS transport implementation in pjproject is broken with inconsistent mutex use both in ioqueue callbacks and ioqueue callers. Unfortunately `PJ_IOQUEUE_CALLBACK_NO_LOCK` is a brand new development with no stability guarantees.

https://github.com/pjsip/pjproject/issues/4818#issuecomment-4030768355

> I notice that PJ_IOQUEUE_CALLBACK_NO_LOCK is not yet enabled by default in 2.16 and that there are additional upcoming fixes for that feature (https://github.com/pjsip/pjproject/commit/21d651d87276281a1257b1bebebebfb3a89200b8, https://github.com/pjsip/pjproject/commit/79b304fb54a299ca70e9821eb1f972074118eb88) that have not been released yet.
> [...]
> What is your level of confidence with PJ_IOQUEUE_CALLBACK_NO_LOCK and what would be your recommendation for PJSIP 2.16 users? Should we enable the feature flag (missing the unreleased bugfixes), backport those fixes, or should we straight up not use SIPTLS with PJSIP 2.16 until the situation has stabilized somewhat?

Answer:

> My recommendation would be to use 2.16 with the addition of all the patches related to PJ_IOQUEUE_CALLBACK_NO_LOCK. We can't guarantee that it will be perfectly stable, but should you encounter any issues, please report it immediately and we'll try to get it fixed ASAP.

So the pjproject position is that the only viable fix is the `PJ_IOQUEUE_CALLBACK_NO_LOCK` but that requires backporting some "unguaranteed" fixes to 2.16. There's no perfect solution, so to the asterisk maintainers I ask which would be the preferred solution:

* If we backport the new ioqueue patches to 2.16 and enable them by default in asterisk, the asterisk project shoulders the burden of guaranteeing stability. These changes affect the ioqueue which impacts every transport (in fact one of the yet-unreleased fixes is for the UDP transport).
  * Is the Asterisk project confident that passing the test suite would be sufficient to validate this brand new `PJ_IOQUEUE_CALLBACK_NO_LOCK` feature?
  * Should additional tests be written to stress-test concurrent access to the transports?
* If we don't do anything, then asterisk 22+ continues to have a worse behavior than asterisk 18: under concurrent writes, TLS connections may fully deadlock instead of resetting.
* Ugly third option: revert https://github.com/pjsip/pjproject/pull/4543 as a patch and live with this "bad record mac" error until upstream stabilizes the feature under the rationale that a transport reset is better than a deadlock. (This is not a personal request, as on our side we will postpone our upgrade to asterisk 22 until this issue is fully sorted out).

**Comment by jcolp** [comment]
Determining the future direction requires deeper research and such. I've raised the internal ticket to a higher priority, but don't anticipate an answer this week.

**Comment by jacotec** [comment]
> * If we don't do anything, then asterisk 22+ continues to have a worse behavior than asterisk 18: under concurrent writes, TLS connections may fully deadlock instead of resetting.

In my opinion this should be avoided at any case. I'm also suffering from TLS reconnections a few times per day, and due to these short dropouts I've changed all to UDP/TCP which is possible - but for some connections TLS is required.

Short dropouts is one thing - **but deadlocks must never happen and should definitely be avoided**. So IMHO either skip that TLS fix fully until `PJ_IOQUEUE_CALLBACK_NO_LOCK` is stable to keep Asterisk itself fine - or with that TLS fix there should be no other way apart from using `PJ_IOQUEUE_CALLBACK_NO_LOCK`.

**Comment by mbradeen** [comment]
@azertyfun - I am able to see the original tls error within a few iterations of flood_ping running on Asterisk 20.7, but so far I haven't seen any indication of a deadlock on the current master branch even after 20+ iterations.  

I've tried running flood_ping both with and without simultaneous outbound tls calls but don't see any calls blocked or OPTIONS fail to send.

How are you detecting the deadlock in this case and what options are you building Asterisk with?

**Comment by azertyfun** [comment]
@mbradeen To reproduce with pjsip directly I had to comment out the "fast track" in `pj_ioqueue_sendto`. Then it was immediately reproducible (see reproduction section of https://github.com/pjsip/pjproject/issues/4818).

Relatedly, I am currently hunting down another issue where we are noticing a condition in our production setup where `ssock->send_buf` fills up for some reason – which as a side-effect makes the ioqueue not go through that "fast track".

We use asterisk 18 with some custom patches. Currently I don't yet have a reason to suspect those being at fault. I reproduced the issue just by running two instances of asterisk in docker, with one endpoint registering with SIPTLS on the other one. Once I have some time I'll look into providing a minimum reproducible setup.

**Comment by azertyfun** [comment]
FYI I am quite confident (though I still need to validate this by pushing a change in production) that a side-effect of this issue is slow transport preventing re-registration due to a separate bug in pjproject:

https://github.com/pjsip/pjproject/issues/4864
https://github.com/pjsip/pjproject/pull/4878

This would pretty much only happen when there is a lot of traffic over a single TLS connection. The symptom is that for 15 minutes after a TLS reconnection, we see very few outgoing registrations succeed. I have a hypothesis for why this happens:

1. TLS disconnection happens
2. https://github.com/pjsip/pjproject/issues/4864 prevents transport destruction
3. ioqueue has a [defense against busy loops](https://github.com/pjsip/pjproject/blob/2.15.1/pjlib/src/pj/ioqueue_epoll.c#L1051) which gets triggered in this case. This **has been confirmed to be happening in production** by taking a runtime trace of asterisk during a coredump:
  ```
    (gdb) bt
  #0  0x00007f776693723f in clock_nanosleep () from /lib/x86_64-linux-gnu/libc.so.6
  #1  0x00007f776693cec7 in nanosleep () from /lib/x86_64-linux-gnu/libc.so.6
  #2  0x00007f776696fa7f in usleep () from /lib/x86_64-linux-gnu/libc.so.6
  #3  0x00007f7767723e49 in pj_thread_sleep (msec=10) at ../src/pj/os_core_unix.c:948
  #4  0x00007f776772262f in pj_ioqueue_poll (ioqueue=0x7f7756581310, timeout=0x7f77543fd4f0) at ../src/pj/ioqueue_epoll.c:1068
  #5  0x00007f7767653e3b in pjsip_endpt_handle_events2 (endpt=0x7f7762731028, max_timeout=0x7f77543fd550, p_count=0x0) at ../src/pjsip/sip_endpoint.c:745
  #6  0x00007f7767653f14 in pjsip_endpt_handle_events (endpt=0x7f7762731028, max_timeout=0x7f77543fd550) at ../src/pjsip/sip_endpoint.c:777
  #7  0x00007f775a94fef3 in monitor_thread_exec (endpt=0x0) at res_pjsip.c:2317
  #8  0x00007f77677238f3 in thread_main (param=0x7f77645670a8) at ../src/pj/os_core_unix.c:701
  #9  0x00007f776726a609 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
  #10 0x00007f7766979353 in clone () from /lib/x86_64-linux-gnu/libc.so.6
  ```

The ongoing _hypothesis_ is that all of this is related and that the `pj_thread_sleep` above is "proof" that we are slowing down ioqueue processing to the point of causing major issues for traffic after a TLS reconnection. However I still cannot fully explain (because I did not dive into the asterisk part of all this) the symptoms as not all TLS connections seem to be impacted as much. Maybe I am missing some nuance.

However we are now going to backport the PR I made in pjproject and hopefully have it in production in the next few weeks. This won't solve the TLS cut itself but will allow us to report back whether we at least are able to always recover quickly.

---

On a broader note, I am disappointed in the correctness and implementation of this part of pjproject. The code is quite complex to grasp, IMO overly so, and I am definitely not a fan of the patterns used to manage the lifetime of a `tdata` which have clearly failed to uphold the implicit internal contract of always invoking the callback.

Furthermore the fact that the ioqueue is asynchronous but the responsibility of calling the callback may fall upon the callee or the caller depending on the "busy-ness" of the system is a massive footgun since it prevents implementation issues from being caught unless actively looked for.

I'm not quite sure to make of all that. Mainly, "it is what it is". But maybe asterisk maintainers ought to pay some attention to unit testing transport logic, especially under load.

**Comment by azertyfun** [comment]
@mbradeen The following PRs should address all remaining concerns with some refactors for the SSL implementation of pjproject:

1. https://github.com/pjsip/pjproject/pull/4878
2. https://github.com/pjsip/pjproject/pull/4909
3. https://github.com/pjsip/pjproject/pull/4910 (currently WIP, I did not yet review it)

Once the last one gets merged, all the general concerns I had about the robustness of the implementation will have been addressed. It also does away with the risky locking mechanisms which is the proper fix for this issue.

Given the extent of the issues being addressed and the fact that none were detected by asterisk's test suite, I wonder if it might be worth your time to create some targeted tests for SIP-TLS to verify that the 3 refactors above do not introduce any regression. In any case I would also encourage the Asterisk team to review the changes.

**Comment by mbradeen** [comment]
@azertyfun - I have tested locally with the patches recommended from the pjproject team and haven't run into any issues this far with them.  However, based on the number of patches required and the number of unrelated patches we added recently, we may end up waiting for the 2.17 release which should cover all.  Either way, thank you for the testsuite suggestion, we will see if any additional coverage is required.

**Comment by mbradeen** [comment]
This issue should be resolved with the merge of https://github.com/asterisk/asterisk/pull/1889
